Main.cpp

```diff
@@ -191,43 +239,63 @@ int main() {
       // canBus.printRxAll();
     }
 
-    chThdSleepMilliseconds(50);
+    chThdSleepMilliseconds(10000);
   }
 }
 
-// module utility functions
 // @brief Convert bytes in RX buffer from ADC reading to temperature
 // @return temp Temperature in degrees C, resolution of .25
+// TODO: Add in per-channel moving average
 uint8_t adc_to_temp(uint8_t* rxbuf) {
   static uint16_t raw{};
-  static int norm{};
+  static double voltage{};
+  static int lookup_index{};
+  static uint8_t* interpolation_segment{};
+  static float interpolated_base_offset{};
+  static uint16_t temp{};
+  static uint8_t compressed_temp{};
+
   // voltage at max supported temperature (63.75C)
-  static constexpr float kMaxTempVoltage = 1.4875;
-  static constexpr float kMinTempVoltage = 2.17;
-  static constexpr float kVRef = 3.316;
-  static constexpr float kAdcMax = 1023.0;
-  // true resolution of ~ 211.5 values across 0-63.75C
-  static constexpr float kMaxTempValue =
-      (kMaxTempVoltage / kVRef) * kAdcMax;  // ~ 461.125
-  static constexpr float kMinTempValue =
-      (kMinTempVoltage / kVRef) * kAdcMax;  // ~ 672.7
-  static constexpr float kMaxIntegerValue =
-      kMinTempValue - kMaxTempValue;  // ~ 211.575
-  // unpack raw reading to continuous 10 bits
-  // A) ((lower 7 of byte 0) << 3) | (lower 3 of byte 1)
-  // B) (upper 7 bits of the reading, in the upper 7 position) |
-  //      (lower 3 bits of the second byte, in the lower 3 position)
-  //
-  raw = ((rxbuf[0] & 0x7f) << 3) | (rxbuf[1] & 0x7);
-  // normalize to 0=63.75C to 255=0C
-  // (decode to degree C by dividing by 4...bit shifting right by 2)
-  norm = ((raw - kMaxTempValue) / kMaxIntegerValue) * 255;
-  // top-out temp at 63.75C and bottom-out at 0C
-  if (norm < 0) {
-    norm = 0;
-  } else if (norm > 255) {
-    norm = 255;
+  static constexpr float kVref = 3.315;
+  static constexpr double kAdcMax = 1023.0;
+  static constexpr uint8_t kMaxTemp = 255;  // max of 63.75C
+
+  // Pack bits from SPI bytes into single 10b ADC reading:
+  // (upper 7 bits of the first byte, in the upper 7 position) |
+  //    (upper 3 bits of the second byte, in the lower 3 position)
+  raw = ((rxbuf[0] & 0x7f) << 3) | ((rxbuf[1] & 0xe0) >> 5);
+
+  // convert raw reading to voltage based on ADC resolution and Vref
+  voltage = (raw / kAdcMax) * kVref;
+
+  // Convert fp voltage to integer (with 0.01V resolution), then normalize to
+  // 0=148 (1.48V) and 69=217 (2.17V)
+  lookup_index = (voltage * 100) - 148;
+
+  // clamp index between 0 and 69
+  // TODO: Try std::clamp (-std=c++1z didn't work)
+  if (lookup_index < 0) {
```

Comment:
The error message said the type deduction was failing. `std::clamp()` uses the same type for all three arguments. Replace this with `lookupIndex = std::clamp<uint8_t>(lookupIndex, 0, g_kMaxVoltageIndex);`. The explicit type is needed here because 0 deduces to `int` otherwise.